packages/core/src/services/chains/ProviderProcessor/index.ts

```diff
@@ -1,148 +1,97 @@
-
-    return Result.ok({ ...response, providerLog })
-  }
+    // AI Provider Data
+    workspaceId: workspace.id,
+    source: source,
+    providerId: apiProvider.id,
+    providerType: apiProvider.provider,
+    // FIXME: This should be polymorphic
+    // https://github.com/latitude-dev/latitude-llm/issues/229
+    documentLogUuid: errorableUuid,
 
-  private async processResponse({
-    aiResult,
-    startTime,
-  }: {
-    aiResult: Awaited<AIReturn<StreamType>>
-    startTime: number
-  }) {
-    const commonData = await this.buildCommonData({ aiResult, startTime })
-
-    if (aiResult.type === 'text') {
-      return processStreamText({ aiResult, commonData })
-    }
-
-    return processStreamObject({ aiResult, commonData })
+    // AI
+    duration,
+    generatedAt: new Date(),
+    model: config.model,
+    config: config,
+    messages: messages,
+    usage: await aiResult.data.usage,
   }
+}
 
-  private async buildCommonData({
+/**
+ * This function is responsible for processing the AI response
+ */
+export async function processResponse({
+  aiResult,
+  startTime,
+  workspace,
+  source,
+  apiProvider,
+  config,
+  messages,
+  errorableUuid,
+}: {
+  aiResult: Awaited<AIReturn<StreamType>>
+  startTime: number
+  workspace: Workspace
+  source: LogSources
+  apiProvider: ProviderApiKey
+  config: PartialConfig
+  messages: Message[]
+  errorableUuid?: string
+}): Promise<ChainStepResponse<StreamType>> {
+  const commonData = await buildCommonData({
     aiResult,
     startTime,
-  }: {
-    aiResult: Awaited<AIReturn<StreamType>>
-    startTime: number
-  }): Promise<StreamCommonData> {
-    const endTime = Date.now()
-    const duration = endTime - startTime
-    return {
-      uuid: generateUUIDIdentifier(),
-
-      // AI Provider Data
-      workspaceId: this.workspace.id,
-      source: this.source,
-      providerId: this.apiProvider.id,
-      providerType: this.apiProvider.provider,
-      // FIXME: This should be polymorphic
-      // https://github.com/latitude-dev/latitude-llm/issues/229
-      documentLogUuid: this.errorableUuid,
+    workspace,
+    source,
+    apiProvider,
+    config,
+    messages,
+    errorableUuid,
+  })
 
-      // AI
-      duration,
-      generatedAt: new Date(),
-      model: this.config.model,
-      config: this.config,
-      messages: this.messages,
-      usage: await aiResult.data.usage,
-    }
+  if (aiResult.type === 'text') {
+    return processStreamText({ aiResult, commonData })
   }
 
-  private checkValidType(aiResult: AIReturn<StreamType>) {
-    const { type } = aiResult
-    const invalidType = type !== 'text' && type !== 'object'
-    if (!invalidType) return Result.nil()
-
-    return Result.error(
-      new ChainError({
-        code: RunErrorCodes.UnsupportedProviderResponseTypeError,
-        message: `Invalid stream type ${type} result is not a textStream or objectStream`,
-      }),
-    )
-  }
+  return processStreamObject({ aiResult, commonData })
 }
```

Comment:
moved to functional implementation and removed the code to create provider log, it was coupled to this service